// github-actions/lib/cli.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseArgs = exports.getHelp = void 0;
const command_line_args_1 = __importDefault(require("command-line-args"));
const command_line_usage_1 = __importDefault(require("command-line-usage"));
const constants_1 = require("./constants");
const indexes_1 = require("./indexes");
function getHelp() {
    return command_line_usage_1.default([
        {
            header: "GitHub Actions Manager",
            content: [
                "CLI tool to install and update and share GitHub Actions workflows",
                "",
                `Documentation: ${constants_1.DOCS_URL}`,
                "",
                "{bold ghactions} - Run in interactive mode",
                "{bold ghactions -i node} - Manage workflows for Node.js projects",
                "{bold ghactions -i python} - Manage workflows for Python projects"
            ].join("\n")
        },
        {
            header: "Options",
            optionList: [
                {
                    name: "names",
                    alias: "n",
                    typeLabel: "name",
                    multiple: true,
                    description: `Workflow name {bold <name>}.yml, {bold all}, or {bold installed}, default: {bold interactive mode}`
                },
                {
                    name: "index",
                    alias: "i",
                    typeLabel: "url",
                    description: "Link to workflows directory {bold .github/workflows}",
                    type: String
                },
                {
                    name: "ref",
                    alias: "r",
                    typeLabel: "version",
                    description: "Update workflows to a specific tag/branch"
                },
                {
                    name: "path",
                    alias: "p",
                    typeLabel: "path",
                    description: `Path to GitHub project root, default {bold CWD}`,
                    type: Boolean
                },
                {
                    name: "update",
                    alias: "u",
                    description: "Apply suggested changes",
                    type: Boolean
                },
                {
                    name: "force",
                    alias: "f",
                    description: "Update user-managed workflow parts",
                    type: Boolean
                },
                {
                    name: "list",
                    alias: "l",
                    description: "List available workflows",
                    type: Boolean
                },
                {
                    name: "diff",
                    alias: "d",
                    description: "Show diff for update and check runs",
                    type: Boolean
                },
                {
                    name: "clean",
                    alias: "c",
                    description: "Do not add {bold github-actions-managed} marker to workflow steps",
                    type: Boolean
                },
                {
                    name: "help",
                    alias: "h",
                    description: "Print this usage guide",
                    type: Boolean
                },
                {
                    name: "version",
                    alias: "v",
                    description: "Show package version",
                    type: Boolean
                }
            ]
        }
    ]);
}
exports.getHelp = getHelp;
function parseArgs() {
    const result = command_line_args_1.default([
        { name: "version", alias: "v", type: Boolean },
        { name: "list", alias: "l", type: Boolean },
        { name: "diff", alias: "d", type: Boolean },
        { name: "clean", alias: "c", type: Boolean },
        { name: "help", alias: "h", type: Boolean },
        { name: "force", alias: "f", type: Boolean },
        { name: "update", alias: "u", type: Boolean },
        { name: "index", alias: "i", type: String },
        { name: "path", alias: "p", type: String },
        {
            name: "names",
            alias: "n",
            type: String,
            multiple: true,
            defaultOption: true
        },
        {
            name: "ref",
            alias: "r",
            type: String
        }
    ]);
    result.names = result.names || [];
    result.help = result.help || false;
    result.force = result.force || false;
    result.update = result.update || false;
    result.ref = result.ref || "";
    result.path = result.path || ".";
    result.index = indexes_1.getIndexURL(result.index || "");
    return result;
}
exports.parseArgs = parseArgs;
