// github-actions/lib/indexes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getShortcut = exports.getIndexResource = exports.getIndexURL = exports.INDEXES = exports.JS_INDEX_URL = void 0;
exports.JS_INDEX_URL = "https://github.com/vemel/github_actions_js/tree/main/nodejs_workflows";
exports.INDEXES = [
    {
        url: exports.JS_INDEX_URL,
        title: "Node.js",
        shortcut: "node"
    },
    {
        url: "https://github.com/vemel/github_actions_js/tree/main/python_workflows",
        title: "Python",
        shortcut: "python"
    },
    {
        url: "https://github.com/actions-rs/example/tree/master/.github/workflows",
        title: "Rust",
        shortcut: "rust"
    },
    {
        url: "https://github.com/mvdan/github-actions-golang/tree/master/.github/workflows",
        title: "Go",
        shortcut: "go"
    },
    {
        url: "https://github.com/dflook/terraform-github-actions/tree/master/example_workflows",
        title: "Terraform",
        shortcut: "terraform"
    },
    {
        url: "https://github.com/julia-actions/Example.jl/tree/action-tests/.github/workflows",
        title: "Julia",
        shortcut: "julia"
    }
];
function getIndexURL(url) {
    for (const index of exports.INDEXES) {
        if (index.url === url || index.shortcut === url)
            return index.url;
    }
    return url;
}
exports.getIndexURL = getIndexURL;
function getIndexResource(url) {
    for (const index of exports.INDEXES) {
        if (index.url === url || index.shortcut === url)
            return index;
    }
    return null;
}
exports.getIndexResource = getIndexResource;
function getShortcut(url) {
    for (const index of exports.INDEXES) {
        if (index.url === url || index.shortcut === url)
            return index.shortcut;
    }
    return null;
}
exports.getShortcut = getShortcut;
