// qiwi-sdk/dist/apis/wallet/webhooks.api.js
"use strict";
const url = require("../shared/url.js"),
  hmac = require("../shared/hmac.js"),
  get = require("../shared/get.js"),
  api = require("./api.js"),
  decode = require("../shared/platform/decode/decode.js");
/**
 * # Уведомления (вебхуки)
 *
 * Хуки или уведомления с данными о событии (платеже/пополнении)
 * отправляются на ваш сервер. В настоящее время поддерживаются
 * только вебхуки (webhook) - сообщения, адресованные веб-сервисам.
 * Для приема вебхуков вам необходимо настроить свой сервер на
 * прием и обработку POST-запросов ([Формат запросов](https://developer.qiwi.com/ru/qiwi-wallet-personal/#hook_format)).
 *
 * **От вашего сервера успешный ответ 200 OK на входящий запрос
 * должен поступить в течение 1-2 сек. Не дождавшись ответа, сервис
 * КИВИ отправляет еще одно уведомление через 10 минут, потом еще
 * одно через 1 час.**
 *
 * Пулы IP-адресов, с которых сервисы QIWI отправляют webhook:
 * - `79.142.16.0/20`
 * - `195.189.100.0/22`
 * - `91.232.230.0/23`
 * - `91.213.51.0/24`
 *
 * Если ваш сервер обработки вебхуков работает за брандмауэром, необходимо добавить эти IP-адреса в список разрешенных адресов входящих TCP-пакетов.
 *
 * @export
 * @class WalletWebhooksApi
 * @extends {WalletApi}
 */
class WalletWebhooksApi extends api.WalletApi {
  constructor() {
    super(...arguments);
    this.keys = new Map();
  }
  /**
   *
   *
   * @protected
   * @return {string}
   * @memberof WalletWebhooksApi
   */
  _getDefaultHookId() {
    if (this.activeId) return this.activeId;
    throw new Error("Unable to get default hook id");
  }
  /**
   * Регистрирует обработчик вебхука
   * @param {string} parameter Адрес сервера обработки вебхуков. **Внимание! Длина исходного (не URL-encoded) адреса сервиса обработчика не должна превышать 100 символов.**
   * @param {number} txnType Тип транзакций, по которым будут включены уведомления.. 0 - "входящие", 1 - "исходящие". 2 - "все"
   * @return {Promise<WebHookInfo>}
   */
  async add(parameter, txnType) {
    const hookResponse = await this.http.put(
      url.url`payment-notifier/v1/hooks`({
        hookType: 1,
        param: parameter,
        txnType
      })
    );
    this.activeId = hookResponse.hookId;
    return hookResponse;
  }
  /**
   * Удаляет обработчик вебхука
   * @param {string} [hookId] UUID вебхука
   * @return {Promise<*>}
   */
  remove(hookId = this._getDefaultHookId()) {
    this.keys.delete(hookId);
    this.activeId = undefined;
    return this.http.delete(url.url`payment-notifier/v1/hooks/${hookId}`());
  }
  /**
   * Получает секретный ключ вебхука
   * @param {string} hookId UUID вебхука
   * @return {Promise<string>}
   */
  async getSecret(hookId = this._getDefaultHookId()) {
    const { key } = await this.http.get(
      url.url`payment-notifier/v1/hooks/${hookId}/key`()
    );
    this.keys.set(hookId, key);
    return key;
  }
  /**
   * Изменяет секретный ключ вебхука
   * @param {string} hookId UUID вебхука
   * @return {Promise<string>}
   */
  async updateSecret(hookId = this._getDefaultHookId()) {
    const { key } = await this.http.post(
      url.url`payment-notifier/v1/hooks/${hookId}/newkey`()
    );
    this.keys.set(hookId, key);
    return key;
  }
  /**
   *
   * @return {Promise<WebHookInfo>}
   */
  async getActiveWebhook() {
    const hookResponse = await this.http.get(
      url.url`payment-notifier/v1/hooks/active`()
    );
    this.activeId = hookResponse.hookId;
    return hookResponse;
  }
  /**
   * Отправляет тестовое уведомление
   * @link https://developer.qiwi.com/ru/qiwi-wallet-personal/#hook_test
   *
   * @return {Promise<*>}
   */
  testActiveWebhook() {
    return this.http.get(url.url`payment-notifier/v1/hooks/test`());
  }
  /**
   * Проверяет подпись уведомления вебхука
   * @param {WebhookTransaction} transaction Объект уведомления транзакции вебхука
   * @return {Promise<boolean | undefined>}
   */
  async checkSign(transaction) {
    var _a;
    const { hookId, payment, hash } = transaction;
    if (!this.keys.has(hookId)) {
      try {
        await this.getSecret(hookId);
      } catch (_b) {
        return;
      }
    }
    const signPayload = payment.signFields.split(",").map((fieldName) => {
      var _a;
      return String(
        (_a = get.getOwnPropertyDeep(payment, fieldName)) !== null && _a !== void 0
          ? _a
          : "null"
      );
    });
    return hmac.compareQiwiHmac(
      decode.base64ToUint8Array(
        (_a = this.keys.get(hookId)) !== null && _a !== void 0 ? _a : ""
      ),
      hash,
      signPayload
    );
  }
}
exports.WalletWebhooksApi = WalletWebhooksApi;
