// qiwi-sdk/dist/apis/shared/index.js
"use strict";
const collect = require("./collect.js"),
  querystring = require("./querystring.js"),
  time = require("./time.js"),
  uuid = require("./uuid.js"),
  hmac = require("./hmac.js"),
  environment = require("./environment.js"),
  http = require("./http.js"),
  http_types = require("./http.types.js"),
  identity = require("./identity.js"),
  url = require("./url.js"),
  urlGlobals = require("./url-globals.js");
exports.collect = collect.collect;
exports.formatQuerystring = querystring.formatQuerystring;
Object.defineProperty(exports, "TimeSpan", {
  enumerable: true,
  get: () => time.TimeSpan
});
exports.TimeSpanMapping = time.TimeSpanMapping;
exports.formatAltLifetimeDate = time.formatAltLifetimeDate;
exports.formatDate = time.formatDate;
exports.formatOffsetAltLifetimeDate = time.formatOffsetAltLifetimeDate;
exports.formatOffsetDate = time.formatOffsetDate;
exports.generateUUID = uuid.generateUUID;
exports.compareHmac = hmac.compareHmac;
exports.compareQiwiHmac = hmac.compareQiwiHmac;
exports.environment = environment.environment;
exports.SimpleJsonHttp = http.SimpleJsonHttp;
exports.HttpError = http_types.HttpError;
exports.USER_AGENT = identity.USER_AGENT;
exports.url = url.url;
exports.URL = urlGlobals.URL;
exports.URLSearchParameters = urlGlobals.URLSearchParameters;
