// qiwi-sdk/dist/apis/shared/time.mjs
import { getOwnProperty } from "./get.mjs";
/**
 * Форматирует дату в понятную для QIWI строку:
 *
 * `ГГГГ-ММ-ДДTЧЧ:ММ:СС+\-ЧЧ:ММ`
 *
 * @param {DateValue} dateTime Аргумент для конструктора
 * @return {string}
 */
function formatDate(dateTime) {
  const date = new Date(dateTime);
  const base = date.toISOString().split(".")[0];
  return `${base}+00:00`;
}
/**
 * **Используется только для параметра `lifetime` при создании ссылки на форму оплаты**
 *
 * Форматирует дату в понятную для QIWI строку:
 *
 * `ГГГГ-ММ-ДДTччмм`
 *
 * @export
 * @param {DateValue} dateTime
 * @return {string} string
 */
function formatAltLifetimeDate(dateTime) {
  const date = new Date(dateTime);
  const base = date.toISOString().split("T")[0];
  const time =
    date.getHours().toString().padStart(2, "0") +
    date.getMinutes().toString().padStart(2, "0");
  return `${base}T${time}`;
}
var TimeSpan;
(function (TimeSpan) {
  TimeSpan[(TimeSpan["Millisecond"] = 1)] = "Millisecond";
  TimeSpan[(TimeSpan["Second"] = 1000)] = "Second";
  TimeSpan[(TimeSpan["Minute"] = 60000)] = "Minute";
  TimeSpan[(TimeSpan["Hour"] = 3600000)] = "Hour";
  TimeSpan[(TimeSpan["Day"] = 86400000)] = "Day";
  TimeSpan[(TimeSpan["Week"] = 604800000)] = "Week";
  TimeSpan[(TimeSpan["Month"] = 2592000000)] = "Month";
  TimeSpan[(TimeSpan["Year"] = 31536000000)] = "Year";
})(TimeSpan || (TimeSpan = {}));
const TimeSpanMapping = Object.freeze({
  ms: TimeSpan.Millisecond,
  msec: TimeSpan.Millisecond,
  millisecond: TimeSpan.Millisecond,
  s: TimeSpan.Second,
  sec: TimeSpan.Second,
  second: TimeSpan.Second,
  m: TimeSpan.Minute,
  min: TimeSpan.Minute,
  minute: TimeSpan.Minute,
  h: TimeSpan.Hour,
  hr: TimeSpan.Hour,
  hour: TimeSpan.Hour,
  d: TimeSpan.Day,
  day: TimeSpan.Day,
  w: TimeSpan.Week,
  week: TimeSpan.Week,
  mn: TimeSpan.Month,
  mon: TimeSpan.Month,
  month: TimeSpan.Month,
  y: TimeSpan.Year,
  yr: TimeSpan.Year,
  year: TimeSpan.Year
});
/**
 *
 *
 * @param {number} amount
 * @param {TimeUnit} unit
 * @param {Date} currentDate
 * @return {Date}
 */
function offsetDate(amount, unit, currentDate) {
  const date = new Date(currentDate);
  if (typeof unit !== "number") {
    unit = getOwnProperty(TimeSpanMapping, unit);
  }
  const time = Math.round(date.getTime() + amount * unit);
  date.setTime(time);
  return date;
}
/**
 * **Используется только для параметра `lifetime` при создании ссылки на форму оплаты**
 *
 * Форматирует дату в понятную для QIWI строку:
 *
 * `ГГГГ-ММ-ДДTччмм`
 *
 * @export
 * @param {number} amount
 * @param {(TimeSpan | number | TimeSpanKeys)} unit
 * @param {Date} [currentDate=new Date()]
 * @return {string} string
 */
function formatOffsetDate(
  amount,
  unit = TimeSpan.Millisecond,
  currentDate = new Date()
) {
  return formatDate(offsetDate(amount, unit, currentDate));
}
/**
 *
 *
 * @export
 * @param {number} amount
 * @param {(TimeSpan | number | TimeSpanKeys)} unit
 * @param {Date} [currentDate=new Date()]
 * @return {string} string
 */
function formatOffsetAltLifetimeDate(
  amount,
  unit = TimeSpan.Millisecond,
  currentDate = new Date()
) {
  return formatAltLifetimeDate(offsetDate(amount, unit, currentDate));
}
export {
  TimeSpan,
  TimeSpanMapping,
  formatAltLifetimeDate,
  formatDate,
  formatOffsetAltLifetimeDate,
  formatOffsetDate
};
