// qiwi-sdk/dist/apis/detector/detector.mjs
import { ApiClass } from "../api.mjs";
import { SimpleJsonHttp } from "../shared/http.mjs";
import { url } from "../shared/url.mjs";
import { formatQuerystring } from "../shared/querystring.mjs";
import { USER_AGENT } from "../shared/identity.mjs";
import { DetectorDetectApi } from "./detect.api.mjs";
/**
 * API получения ID провайдера QIWI по Номеру Телефона/Карте
 *
 * @export
 * @class Detector
 */
class Detector extends ApiClass {
  /**
   *
   *
   * @static
   * @return {Detector} Detector
   * @memberof Detector
   */
  static create() {
    return new this({ http: this.httpClientFactory() });
  }
  /**
   * Creates an instance of P2p.
   * @param {P2pApiOptions} [options]
   * @memberof P2p
   */
  constructor({ http = Detector.httpClientFactory() } = {}) {
    super({ http });
    this.detectProvider = new Detector.DetectApi(this._options);
  }
}
Detector.DetectApi = DetectorDetectApi;
/**
 *
 *
 * @static
 * @return {SimpleJsonHttp} SimpleJsonHttp
 * @memberof Detector
 */
Detector.httpClientFactory = () => {
  const http = new SimpleJsonHttp();
  http.client.options = {
    ...http.client.options,
    stringifyBody: (body) => formatQuerystring(body),
    baseURL: url`https://qiwi.com/`(),
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Accept: "application/json",
      "User-Agent": USER_AGENT
    },
    okStatusCodes: [200],
    timeout: 10000
  };
  return http;
};
export { Detector };
