// qiwi-sdk/dist/apis/p2p/p2p.js
"use strict";
const api = require("../api.js"),
  environment = require("../shared/environment.js"),
  http = require("../shared/http.js"),
  url = require("../shared/url.js"),
  identity = require("../shared/identity.js"),
  bills_api = require("./bills.api.js"),
  p2p_errors = require("./p2p.errors.js"),
  p2p_middleware = require("./p2p.middleware.js"),
  p2p_types = require("./p2p.types.js");
/**
 * # P2P-счета
 * [Документация QIWI](https://developer.qiwi.com/ru/p2p-payments/)
 *
 * @export
 * @class P2p
 */
class P2p extends api.ApiClass {
  /**
   *
   *
   * @static
   * @param {string} secretKey
   * @param {string} [publicKey=""]
   * @return {P2p} P2p
   * @memberof P2p
   */
  static create(secretKey, publicKey = "") {
    return new this({
      http: this.httpClientFactory(secretKey),
      publicKey,
      secretKey
    });
  }
  /**
   *
   *
   * @static
   * @param {string} [secretKey=process.env.QIWI_SECRET_KEY]
   * @param {string} [publicKey=process.env.QIWI_PUBLIC_KEY]
   * @return {P2p} P2p
   * @memberof P2p
   */
  static env(
    secretKey = environment.environment.QIWI_SECRET_KEY,
    publicKey = environment.environment.QIWI_PUBLIC_KEY
  ) {
    return this.create(secretKey, publicKey);
  }
  /**
   * Creates an instance of P2p.
   * @param {P2pApiOptions} [options]
   * @memberof P2p
   */
  constructor({
    publicKey = "",
    secretKey = "",
    http = P2p.httpClientFactory(secretKey)
  } = {}) {
    super({ publicKey, secretKey, http });
    this.bills = new P2p.BillsApi(this._options);
  }
  /**
   * `[Экспериментально]` Упрощает интеграцию с `express`
   *
   * ## Это middleware кидает ошибки, позаботьтесь об их обработке
   *
   * @param {Object} [options] Параметры обработки запроса
   * @param {boolean} [options.memo=true] Флаг для включения/отключения пропуска повторяющихся запросов, если один из них был успешно обработан
   *
   * @param {RequestHandler<Record<string, string>, any, BillStatusNotificationBody>} handler
   *
   * @return {RequestHandler}
   *
   * ##### Пример:
   * **В начале файла**
   * ```js
   * const p2p = new QIWI.P2P(process.env.QIWI_PRIVATE_KEY);
   * ```
   * *`Вариант 1 - Классический`*
   *
   * ```js
   * app.post('/webhook/qiwi', p2p.notificationMiddleware(), (req, res) => {
   *  req.body // Это `BillStatusNotificationBody`
   * })
   * ```
   *
   * *`Вариант 2 - Если нужны подсказки типов`*
   *
   * ```js
   * app.post('/webhook/qiwi', p2p.notificationMiddleware({}, (req, res) => {
   *  req.body // Это `BillStatusNotificationBody`
   * }))
   * ```
   *
   * **Обработка ошибок**
   * ```js
   * app.use((error, request, response, next) => {
   *  console.log(error); // [Error: Notification signature mismatch]
   * })
   * ```
   */
  notificationMiddleware(options = {}, handler) {
    return p2p_middleware.createP2pNotificationMiddleware(
      this.bills,
      options,
      handler
    );
  }
}
P2p.BillsApi = bills_api.P2pBillsApi;
P2p.BillCurrency = p2p_types.BillCurrency;
P2p.BillPaySource = p2p_types.BillPaySource;
P2p.BillStatus = p2p_types.BillStatus;
P2p.Currency = p2p_types.BillCurrency;
P2p.PaySource = p2p_types.BillPaySource;
P2p.Status = p2p_types.BillStatus;
/**
 *
 *
 * @static
 * @param {string} secretKey
 * @return {SimpleJsonHttp} SimpleJsonHttp
 * @memberof P2p
 */
P2p.httpClientFactory = (secretKey) => {
  const http$1 = new http.SimpleJsonHttp();
  http$1.client.options = {
    ...http$1.client.options,
    baseURL: url.url`https://api.qiwi.com/partner/bill/v1/bills/`(),
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      "User-Agent": identity.USER_AGENT,
      Authorization: `Bearer ${secretKey}`
    },
    okStatusCodes: [200],
    timeout: 10000,
    mapHttpErrors: p2p_errors.mapHttpErrors
  };
  return http$1;
};
exports.P2p = P2p;
