// qiwi-sdk/dist/apis/error.mjs
/**
 *
 *
 * @class ErrorWithCause
 * @extends {Error}
 */
class ErrorWithCause extends Error {
  /**
   * Creates an instance of ErrorWithCause.
   * @param {string} message
   * @param {Error} [cause]
   * @memberof ErrorWithCause
   */
  constructor(message, cause) {
    const parameters = [message, { cause }];
    super(...parameters);
    this.message = message;
    this.cause = cause;
  }
}
/**
 * Ошибка, которую можно преобразовать в JSON
 *
 * @export
 * @class ExtendedError
 * @extends {Error}
 */
class ExtendedError extends ErrorWithCause {
  constructor() {
    super(...arguments);
    this.name = this.constructor.name;
  }
  /**
   *
   *
   * @return {*} Error
   * @memberof ExtendedError
   */
  toJSON() {
    return {
      message: this.message,
      name: this.name,
      stack: this.stack
    };
  }
  /**
   *
   *
   * @readonly
   * @type {string}
   * @memberof ExtendedError
   */
  get [Symbol.toStringTag]() {
    return this.name;
  }
}
export { ErrorWithCause, ExtendedError };
